homeassistant/config_entries.py

```diff
@@ -750,30 +750,51 @@ def async_update_entry(
         data: dict = _UNDEF,
         options: dict = _UNDEF,
         system_options: dict = _UNDEF,
-    ) -> None:
-        """Update a config entry."""
-        if unique_id is not _UNDEF:
+    ) -> bool:
+        """Update a config entry.
+
+        If the entry was changed, the update_listeners are
+        fired and this function returns True
+
+        If the entry was not changed, the update_listeners are
+        not fired and this function returns False
+        """
+        changed = False
+
+        if unique_id is not _UNDEF and entry.unique_id != unique_id:
+            changed = True
             entry.unique_id = cast(Optional[str], unique_id)
 
-        if title is not _UNDEF:
+        if title is not _UNDEF and entry.title != title:
+            changed = True
             entry.title = cast(str, title)
 
-        if data is not _UNDEF:
+        if data is not _UNDEF and dict(entry.data) != dict(data):
+            changed = True
             entry.data = MappingProxyType(data)
 
-        if options is not _UNDEF:
+        if options is not _UNDEF and dict(entry.options) != dict(options):
```

Comment:
Why do we need to copy `options` or `data` that is passed in?

Reply:
Actually don't need to copy anything.

```
Python 3.8.5 (default, Jul 29 2020, 16:23:57)
[Clang 11.0.3 (clang-1103.0.32.62)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from types import MappingProxyType
>>> MappingProxyType({"a":2}) == {"a":2}
True
```
```suggestion
        if options is not _UNDEF and entry.options != options:
```